[ThinLTO] Shrink GlobalValueSummary by 8 bytes

kazutakahirata opened the pull request:
During the ThinLTO indexing step for one of our large applications, we
create 7.5 million instances of GlobalValueSummary.

Changing:

  std::vector<ValueInfo> RefEdgeList;

to:

  SmallVector<ValueInfo, 0> RefEdgeList;

in GlobalValueSummary reduces the size of each instance by 8 bytes.
The rest of the patch makes the same change to other places so that
the types stay compatible across function boundaries.

kazutakahirata committed:
[ThinLTO] Shrink GlobalValueSummary by 8 bytes

During the ThinLTO indexing step for one of our large applications, we
create 7.5 million instances of GlobalValueSummary.

Changing:

  std::vector<ValueInfo> RefEdgeList;

to:

  SmallVector<ValueInfo, 0> RefEdgeList;

in GlobalValueSummary reduces the size of each instance by 8 bytes.
The rest of the patch makes the same change to other places so that
the types stay compatible across function boundaries.

teresajohnson commented on llvm/lib/Analysis/ModuleSummaryAnalysis.cpp:
```diff
@@ -99,10 +99,11 @@ extern cl::opt<bool> MemProfReportHintedSizes;
-                         bool &RefLocalLinkageIFunc) {
+static bool
+findRefEdges(ModuleSummaryIndex &Index, const User *CurUser,
+             SetVector<ValueInfo, SmallVector<ValueInfo, 0>> &RefEdges,
```
According to https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h (see the "Note" box):

```
Prefer to use ArrayRef<T> or SmallVectorImpl<T> as a parameter type.

It’s rarely appropriate to use SmallVector<T, N> as a parameter type. If an API only reads from the vector, it should use ArrayRef.
```

I see in some places you have used SmallVectorImpl for the parameter but others like here just SmallVector. These should presumably all be either SmallVectorImpl or ArrayRef.

kazutakahirata replied:
Thanks!  I'll go through the parameter types.  By the way, here, we have `SetVector<ValueInfo, SmallVector<ValueInfo, 0>> &`.  We can cast `SmallVector<T, 0> &` to `SmallVectorImpl<T> &`, but I don't think we can cast `SetVector<ValueInfo, SmallVector<ValueInfo, 0>> &` to `SetVector<ValueInfo, SmallVectorImpl<ValueInfo>> &`.  Note that `SmallVector` is embedded in a parent type, namely `SetVector`.

kazutakahirata committed:
Use SmallVectorImpl

kazutakahirata commented:
Please take a look.  Thanks!

- I've updated the PR to use `SmallVector<ValueInfo> &` and `SmallVector<ValueInfo> &&` where I can.
- I've added a missing `std::move` to `ModuleSummaryIndexYAML.h`.

minglotus-6 commented on llvm/include/llvm/IR/ModuleSummaryIndex.h:
```diff
@@ -538,10 +538,11 @@ class GlobalValueSummary {
   /// from within a function). This does not include functions called, which
   /// are listed in the derived FunctionSummary object.
-  std::vector<ValueInfo> RefEdgeList;
+  SmallVector<ValueInfo, 0> RefEdgeList;
```
nit: add a comment to explain the motivation to use `SmallVector<T, 0>` for future readers / maintainers.

kazutakahirata replied:
Fixed in the latest iteration.  Thanks!

mingmingl-llvm reviewed (approved):
lgtm as long as interface discussion (https://github.com/llvm/llvm-project/pull/107342/files#r1745875758) is resolved

teresajohnson reviewed (approved):
lgtm with mingming's comment suggestion

kazutakahirata committed:
Add a comment about SmallVector.
